File: src/java/nz/co/searchwellington/controllers/models/TagModelBuilder.java
 package nz.co.searchwellington.controllers.models;

import java.util.List;

import javax.servlet.http.HttpServletRequest;

import nz.co.searchwellington.controllers.RelatedTagsService;
import nz.co.searchwellington.controllers.RssUrlBuilder;
import nz.co.searchwellington.controllers.UrlBuilder;
import nz.co.searchwellington.feeds.RssfeedNewsitemService;
import nz.co.searchwellington.model.Feed;
import nz.co.searchwellington.model.FeedNewsitem;
import nz.co.searchwellington.model.Resource;
import nz.co.searchwellington.model.Tag;
import nz.co.searchwellington.model.TagContentCount;
import nz.co.searchwellington.model.Website;
import nz.co.searchwellington.repositories.ConfigDAO;
import nz.co.searchwellington.repositories.ResourceRepository;
import nz.co.searchwellington.utils.UrlFilters;

import org.apache.log4j.Logger;
import org.springframework.web.servlet.ModelAndView;

public class TagModelBuilder extends AbstractModelBuilder implements ModelBuilder {
	
	Logger log = Logger.getLogger(TagModelBuilder.class);
    	
	private ResourceRepository resourceDAO;
	private RssUrlBuilder rssUrlBuilder;
	private UrlBuilder urlBuilder;
	private RelatedTagsService relatedTagsService;
	private ConfigDAO configDAO;
	private RssfeedNewsitemService rssfeedNewsitemService;
	
	 
	public TagModelBuilder(ResourceRepository resourceDAO, RssUrlBuilder rssUrlBuilder, UrlBuilder urlBuilder, RelatedTagsService relatedTagsService, ConfigDAO configDAO, RssfeedNewsitemService rssfeedNewsitemService) {
		this.resourceDAO = resourceDAO;
		this.rssUrlBuilder = rssUrlBuilder;
		this.urlBuilder = urlBuilder;
		this.relatedTagsService = relatedTagsService;
		this.configDAO = configDAO;
		this.rssfeedNewsitemService = rssfeedNewsitemService;
	}

	
	@SuppressWarnings("unchecked")
	public boolean isValid(HttpServletRequest request) {
		List<Tag> tags = (List<Tag>) request.getAttribute("tags");
		boolean isSingleTagPage = tags != null && tags.size() == 1;
		return isSingleTagPage;
	}

	@SuppressWarnings("unchecked")
	public ModelAndView populateContentModel(HttpServletRequest request, boolean showBroken) {
		if (isValid(request)) {
			log.info("Building tag page model");
			List<Tag> tags = (List<Tag>) request.getAttribute("tags");
			Tag tag = tags.get(0);
			return populateTagPageModelAndView(tag, showBroken);
		}
		return null;
	}
	
	@SuppressWarnings("unchecked")
	public void populateExtraModelConent(HttpServletRequest request, boolean showBroken, ModelAndView mv) {
		List<Tag> tags = (List<Tag>) request.getAttribute("tags");
		Tag tag = tags.get(0);
		List<TagContentCount> relatedTagLinks = relatedTagsService.getRelatedTagLinks(tag, showBroken);
		if (relatedTagLinks.size() > 0) {
			mv.addObject("related_tags", relatedTagLinks);
		}
		
		populateCommentedTaggedNewsitems(mv, tag, showBroken);
		mv.addObject("last_changed", resourceDAO.getLastLiveTimeForTag(tag));
		populateRelatedFeed(mv, tag);
		populateGeocoded(mv, showBroken, tag);		
		populateTagFlickrPool(mv, tag);
	}
 	
	

    private void populateTagFlickrPool(ModelAndView mv, Tag tag) {
        if (tag.getFlickrCount() > 0) {
            mv.addObject("flickr_count", tag.getFlickrCount());
            mv.addObject("escaped_flickr_group_id", UrlFilters.encode(configDAO.getFlickrPoolGroupId()));
        }
    }
	
	private ModelAndView populateTagPageModelAndView(Tag tag, boolean showBroken) {		
		ModelAndView mv = new ModelAndView();				
		mv.addObject("tag", tag);
		mv.addObject("heading", tag.getDisplayName());        		
		mv.addObject("description", rssUrlBuilder.getRssDescriptionForTag(tag));
		mv.addObject("link", urlBuilder.getTagUrl(tag));	
		
		final List<Website> taggedWebsites = resourceDAO.getTaggedWebsites(tag, showBroken, MAX_WEBSITES);
		final List<Resource> taggedNewsitems = resourceDAO.getTaggedNewitems(tag, showBroken, MAX_NEWSITEMS);         
		
		mv.addObject("main_content", taggedNewsitems);
		mv.addObject("websites", taggedWebsites);
	
		if (taggedNewsitems.size() > 0) {
			 setRss(mv, rssUrlBuilder.getRssTitleForTag(tag), rssUrlBuilder.getRssUrlForTag(tag));
		}
				
		mv.setViewName("tag");
		return mv;
	}
	
	

    private void populateCommentedTaggedNewsitems(ModelAndView mv, Tag tag, boolean showBroken) {
        List<Resource> allCommentedNewsitems = resourceDAO.getCommentedNewsitemsForTag(tag, showBroken, MAX_NUMBER_OF_COMMENTED_TO_SHOW + 1);        
        List<Resource>commentedToShow;
        if (allCommentedNewsitems.size() <= MAX_NUMBER_OF_COMMENTED_TO_SHOW) {
            commentedToShow = allCommentedNewsitems;            
        } else {
            commentedToShow = allCommentedNewsitems.subList(0, MAX_NUMBER_OF_COMMENTED_TO_SHOW_IN_RHS);
            final String moreCommentsUrl = urlBuilder.getTagCommentUrl(tag);
            mv.addObject("commented_newsitems_moreurl", moreCommentsUrl);
            // TODO count
        }        
        mv.addObject("commented_newsitems", commentedToShow);        
        mv.addObject("tag_watchlist", resourceDAO.getTagWatchlist(tag, showBroken));        
    }
	
    
    private void populateGeocoded(ModelAndView mv, boolean showBroken, Tag tag) {
        List<Resource> geocoded = resourceDAO.getAllValidGeocodedForTag(tag, MAX_NUMBER_OF_GEOTAGGED_TO_SHOW, showBroken);
        log.info("Found " + geocoded.size() + " valid geocoded resources for tag: " + tag.getName());      
        if (geocoded.size() > 0) {
            mv.addObject("geocoded", geocoded);
        }
    }
	
    
    private void populateRelatedFeed(ModelAndView mv, Tag tag) {       
        Feed relatedFeed = tag.getRelatedFeed();
        if (relatedFeed != null) {
            log.info("Related feed is: " + relatedFeed.getName());
            List<FeedNewsitem> relatedFeedItems = rssfeedNewsitemService.getFeedNewsitems(relatedFeed);
            mv.addObject("related_feed", relatedFeed);   
            mv.addObject("related_feed_items", relatedFeedItems);            
        } else {
            log.debug("No related feed.");
        }
    }
    
}

File: src/main/scala/nz/co/searchwellington/controllers/models/helpers/TagModelBuilder.scala
package nz.co.searchwellington.controllers.models.helpers

import io.opentelemetry.api.trace.Span
import nz.co.searchwellington.ReasonableWaits
import nz.co.searchwellington.controllers.admin.AdminUrlBuilder
import nz.co.searchwellington.controllers.{CommonModelObjectsService, RssUrlBuilder}
import nz.co.searchwellington.model.frontend.{Action, FrontendResource}
import nz.co.searchwellington.model.{Resource, Tag, TagArchiveLink, User}
import nz.co.searchwellington.permissions.EditPermissionService
import nz.co.searchwellington.repositories.{ContentRetrievalService, TagDAO}
import nz.co.searchwellington.tagging.RelatedTagsService
import nz.co.searchwellington.urls.UrlBuilder
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import org.springframework.ui.ModelMap

import java.util
import javax.servlet.http.HttpServletRequest
import scala.concurrent.{ExecutionContext, Future}
import scala.jdk.CollectionConverters._

@Component class TagModelBuilder @Autowired()(rssUrlBuilder: RssUrlBuilder, val urlBuilder: UrlBuilder,
                                              relatedTagsService: RelatedTagsService,
                                              val contentRetrievalService: ContentRetrievalService,
                                              commonAttributesModelBuilder: CommonAttributesModelBuilder,
                                              tagDAO: TagDAO,
                                              editPermissionService: EditPermissionService,
                                              adminUrlBuilder: AdminUrlBuilder) extends ModelBuilder
  with CommonSizes with ReasonableWaits with CommonModelObjectsService with ArchiveMonths {

  private val TAG = "tag"
  private val TAGS = "tags"
  private val TAG_WATCHLIST = "tag_watchlist"
  private val TAG_FEEDS = "tag_feeds"
  private val WEBSITES = "websites"

  def isValid(request: HttpServletRequest): Boolean = {
    val tags = request.getAttribute(TAGS).asInstanceOf[Seq[Tag]]
    tags != null && tags.size == 1
  }

  def populateContentModel(request: HttpServletRequest, loggedInUser: Option[User])(implicit ec: ExecutionContext, currentSpan: Span): Future[Option[ModelMap]] = {

    def populateTagPageModelAndView(tag: Tag): Future[Option[ModelMap]] = {
      val eventualTaggedNewsitems = contentRetrievalService.getTaggedNewsitems(tag, MAX_NEWSITEMS, loggedInUser)
      val eventualChildTags = tagDAO.loadTagsByParent(tag._id)
      val eventualMaybeParent = tag.parent.map { pid =>
        tagDAO.loadTagByObjectId(pid)
      }.getOrElse {
        Future.successful(None)
      }

      val eventualMaybeModel = for {
        taggedNewsitemsAndTotalCount <- eventualTaggedNewsitems
        children <- eventualChildTags
        maybeParent <- eventualMaybeParent

      } yield {
        val totalNewsitems = taggedNewsitemsAndTotalCount._2

        val mv = new ModelMap().
          addAttribute(TAG, tag).
          addAttribute("heading", tag.display_name).
          addAttribute("rss_feed_label", tag.display_name.toLowerCase).
          addAttribute("description", rssUrlBuilder.getRssDescriptionForTag(tag)).
          addAttribute("link", urlBuilder.fullyQualified(urlBuilder.getTagUrl(tag))).
          addAttribute("parent", maybeParent.orNull)

        val taggedNewsitems = taggedNewsitemsAndTotalCount._1
        mv.addAttribute(MAIN_CONTENT, taggedNewsitems.asJava)
        mv.addAttribute("main_heading", tag.display_name + " related newsitems")

        setRss(mv)

        if (totalNewsitems > MAX_NEWSITEMS) {
          monthOfLastItem(taggedNewsitems).foreach { i =>
            val link = TagArchiveLink(tag = tag, interval = i, count = None)
            mv.addAttribute("more", urlBuilder.getArchiveLinkUrl(link))
          }
        }

        if (children.nonEmpty) {
          mv.addAttribute("children", children.asJava)
        }

        Some(mv)

      }

      eventualMaybeModel.map { mvo =>
        mvo.map { mv =>
          val actions = if (editPermissionService.canEdit(tag, loggedInUser)) {
            Seq(
              Action("Edit", adminUrlBuilder.getEditTagUrl(tag)),
              Action("Delete", adminUrlBuilder.deleteTagUrl(tag)),
              Action("Autotag", urlBuilder.getAutoTagUrl(tag))
            )
          } else {
            Seq.empty
          }
          mv.addAttribute("actions", actions.asJava)
        }
      }
    }

    populateTagPageModelAndView(tagFromRequest(request))
  }

  def populateExtraModelContent(request: HttpServletRequest, loggedInUser: Option[User])(implicit ec: ExecutionContext, currentSpan: Span): Future[ModelMap] = {
    val tag = tagFromRequest(request)

    val eventualGeotaggedNewsitems = contentRetrievalService.getGeotaggedNewsitemsForTag(tag, MAX_NUMBER_OF_GEOTAGGED_TO_SHOW, loggedInUser = loggedInUser)
    val eventualTaggedWebsites = contentRetrievalService.getTaggedWebsites(tag, MAX_WEBSITES, loggedInUser)
    val eventualRelatedTagLinks = relatedTagsService.getRelatedTagsForTag(tag, 8, loggedInUser)
    val eventualRelatedPublishersForTag = relatedTagsService.getRelatedPublishersForTag(tag, 8, loggedInUser)
    val eventualTagWatchlist = contentRetrievalService.getTagWatchlist(tag, loggedInUser)
    val eventualTagFeeds = contentRetrievalService.getTaggedFeeds(tag, loggedInUser)

    for {
      geotaggedNewsitems <- eventualGeotaggedNewsitems
      taggedWebsites <- eventualTaggedWebsites
      relatedTagLinks <- eventualRelatedTagLinks
      relatedPublishersForTag <- eventualRelatedPublishersForTag
      tagWatchList <- eventualTagWatchlist
      tagFeeds <- eventualTagFeeds
      archiveLinks <- contentRetrievalService.getTagArchiveMonths(tag, loggedInUser)
      latestNewsitems <- latestNewsitems(loggedInUser)

    } yield {
      val mv = new ModelMap()
      mv.addAttribute(WEBSITES, taggedWebsites.asJava)

      if (relatedTagLinks.nonEmpty) {
        mv.addAttribute("related_tags", relatedTagLinks.asJava)
      }
      if (relatedPublishersForTag.nonEmpty) {
        mv.addAttribute("related_publishers", relatedPublishersForTag.asJava)
      }
      if (geotaggedNewsitems.nonEmpty) {
        mv.addAttribute("geocoded", geotaggedNewsitems.asJava)
      }
      if (archiveLinks.nonEmpty) {
        val tagArchiveLinks = archiveLinks.map { a =>
          TagArchiveLink(tag = tag, interval = a.interval, count = a.count)
        }
        mv.addAttribute("archive_links", tagArchiveLinks.asJava)
      }
      mv.addAttribute(TAG_WATCHLIST, tagWatchList.asJava)
      mv.addAttribute(TAG_FEEDS, tagFeeds.asJava)
      mv.addAllAttributes(latestNewsitems)
      mv
    }
  }

  def getViewName(mv: ModelMap, loggedInUser: Option[User]): String = {
    val mainContent = mv.get(MAIN_CONTENT).asInstanceOf[util.List[Resource]]
    val taggedWebsites = mv.get(WEBSITES).asInstanceOf[util.List[Resource]]
    val tagWatchlist = mv.get(TAG_WATCHLIST).asInstanceOf[util.List[Resource]]
    val tagFeeds = mv.get(TAG_FEEDS).asInstanceOf[util.List[Resource]]

    val hasSecondaryContent = !taggedWebsites.isEmpty || !tagWatchlist.isEmpty || !tagFeeds.isEmpty
    val isOneContentType = mainContent.isEmpty || !hasSecondaryContent

    if (isOneContentType) {
      "tagOneContentType"
    } else {
      "tag"
    }
  }

  def setRss(mv: ModelMap): ModelMap = {
    val tag = mv.get(TAG).asInstanceOf[Tag]
    val mainContent = mv.get(MAIN_CONTENT).asInstanceOf[util.List[FrontendResource]].asScala
    if (mainContent.nonEmpty) {
      commonAttributesModelBuilder.setRss(mv, rssUrlBuilder.getRssTitleForTag(tag), rssUrlBuilder.getRssUrlForTag(tag))
    }
    mv
  }

  private def tagFromRequest(request: HttpServletRequest): Tag = {
    val tags = request.getAttribute(TAGS).asInstanceOf[Seq[Tag]]
    tags.head
  }

}
